src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -401,6 +415,212 @@ static void PromoteCommonCommitCharactersOntoList(LSP.VSCompletionList completio
+                }
+                else
+                {
+                    // This is a new completion request, clear out the last result Id for incomplete results.
+                    _lastIncompleteResultId = null;
+                    result = await CalculateListAsync(request, document, position, completionTrigger, completionOptions, completionService, _completionListCache, cancellationToken).ConfigureAwait(false);
+                }
+
+                if (result == null)
+                {
+                    return null;
+                }
+
+                var resultId = result.Value.ResultId;
+                var (completionList, isIncomplete) = FilterCompletionList(result.Value.List, completionListSpan, completionTrigger, sourceText, document);
+
+                // The list is incomplete, update our cached resultId so we can avoid recomputing the list for followup TriggerForIncompleteCompletions triggers.
+                // Even if the current filtered list is complete, the user may delete some filter text causing the client to re-query us for a previous incomplete response.
+                // So we avoid clearing the cached resultId until the client asks us for an entirely new completion list (handled above).
+                if (isIncomplete)
+                {
+                    // Store the source text as what it would be without the current completion list span.  This allows us to easily
+                    // ensure that the source texts match between an incomplete response and a new TriggerForIncompleteCompletion request.
+                    _lastIncompleteResultId = (sourceText.WithChanges(new TextChange(completionListSpan, string.Empty)), resultId);
+                }
+
+                return (completionList, isIncomplete, resultId);
+            }
+        }
+
+        private static async Task<(CompletionList CompletionList, long ResultId)?> GetCompletionListForIncompleteTriggerAsync(
+            LSP.CompletionParams request,
+            (SourceText, long)? lastResultId,
+            TextSpan completionListSpan,
+            Document document,
+            SourceText sourceText,
+            OptionSet completionOptions,
+            CompletionListCache completionListCache,
+            CompletionService completionService,
+            CancellationToken cancellationToken)
+        {
+            // Map back the current source text to what the source text would be without the current completion span.
+            // We use this for two purposes:
+            //   1.  Compare against the source text we stored with the incomplete resultId to ensure we're not using a stale cached list.
+            //   2.  If we do not have a cached list (or it didn't match), we can re-compute the list.
+            var textWithoutCompletionListSpan = sourceText.WithChanges(new TextChange(completionListSpan, string.Empty));
```

Comment:
this seems pretty heavyweight.  is the logic of LSP not that they just ask for more results from the original request?  and you can/should do all computation aagainst the state of the system when the original request came in?

Reply:
per offline conv, this is removed